cloudsql/sqlserver/database-sql/cloudsql.go

```diff
@@ -0,0 +1,306 @@
+// app presented at Cloud Next '19 as seen in this video:
+// https://www.youtube.com/watch?v=qVgzP3PsXFw&t=1833s
+package main
+
+import (
+	"database/sql"
+	"fmt"
+	"html/template"
+	"log"
+	"math"
+	"net/http"
+	"os"
+	"strconv"
+
+	_ "github.com/denisenkom/go-mssqldb"
+)
+
+// parsedTemplate is the global parsed HTML template.
+var parsedTemplate *template.Template
+
+// vote struct contains a single row from the votes table in the database.
+// Each vote includes a candidate ("TABS" or "SPACES") and a timestamp.
+type vote struct {
+	Candidate string
+	VoteTime  sql.NullTime
+}
+
+// voteDiff is used to provide a string representation of the current voting
+// margin, such as "1 vote" (singular) or "2 votes" (plural).
+type voteDiff int
+
+func (v voteDiff) String() string {
+	if v == 1 {
+		return "1 vote"
+	}
+	return strconv.Itoa(int(v)) + " votes"
+}
+
+// templateData struct is used to pass data to the HTML template.
+type templateData struct {
+	TabsCount   uint
+	SpacesCount uint
+	VoteMargin  string
+	RecentVotes []vote
+}
+
+// App struct contains global state.
+type App struct {
+	// db is the global database connection pool.
+	db *sql.DB
+}
+
+// indexHandler handles requests to the / route.
+func (app *App) indexHandler(w http.ResponseWriter, r *http.Request) {
+	switch r.Method {
+	case "GET":
+		if err := showTotals(w, r, app); err != nil {
+			log.Printf("showTotals: %v", err)
+			http.Error(w, "Internal Server Error", http.StatusInternalServerError)
+		}
+	case "POST":
+		if err := saveVote(w, r, app); err != nil {
+			log.Printf("saveVote: %v", err)
+			http.Error(w, "Internal Server Error", http.StatusInternalServerError)
+		}
+	default:
+		http.Error(w, fmt.Sprintf("HTTP Method %s Not Allowed", r.Method), http.StatusMethodNotAllowed)
+	}
+}
+
+func main() {
+
+	var err error
+
+	parsedTemplate, err = template.ParseFiles("templates/index.html")
+	if err != nil {
+		log.Fatalf("unable to parse template file: %s", err)
+	}
+
+	app := &App{}
+
+	// If the optional DB_TCP_HOST environment variable is set, it contains
+	// the IP address and port number of a TCP connection pool to be created,
+	// such as "127.0.0.1:1433". If DB_TCP_HOST is not set, a Unix socket
+	// connection pool will be created instead.
+	if os.Getenv("DB_TCP_HOST") != "" {
+		app.db, err = initTcpConnectionPool()
+		if err != nil {
+			log.Fatalf("initTcpConnectionPool: unable to connect: %s", err)
+		}
+	} else {
+		app.db, err = initSocketConnectionPool()
+		if err != nil {
+			log.Fatalf("initSocketConnectionPool: unable to connect: %s", err)
+		}
+	}
+
+	// Drop the votes table if it already exists.
+	if _, err = app.db.Exec(`DROP TABLE IF EXISTS votes;`); err != nil {
+		log.Fatalf("DB.Exec: unable to drop votes table: %s", err)
+	}
+	// Create the votes table.
+	if _, err = app.db.Exec(`CREATE TABLE votes
+	( vote_id int IDENTITY(1,1) PRIMARY KEY, time_cast DATETIME NOT NULL,
+	candidate CHAR(6) NOT NULL );`); err != nil {
```

Comment:
Might split this up since it's a bit hard to follow the multi-line inline declaration.